File: client/js/inventory.js
define(['exceptions', 'inventoryitem', 'inventoryitemfactory'], function (Exceptions, InventoryItem, InventoryItemFactory) {

  var Inventory = Class.extend({
    init: function (data) {
      this._size = 12;
      this._items = [];

      if (data) {
        this.loadFromObject(data);
      }
    },

    get id() {
      return this._id;
    },

    get size() {
      return this._size;
    },

    add: function (item) {

    },

    remove: function (idx) {
      if (this._items[idx]) {
        InventoryItemFactory.remove(this._items[idx].id);
        this._items[idx] = null;
      }
    },

    throwItem: function (idx) {
      var item = this._items[idx];
      if (item) {
        this.remove(idx);
        globalGame.client.sendThrowItem(item);
      }
    },

    find: function (itemId) {
      for (var i in this._items) {
        var item = this._items[i];
        if (item && item.id == itemId) {
          return item;
        }
      }

      return null;
    },

    swap: function (first, second) {
      var temp = this._items[first];
      this._items[first] = this._items[second];
      this._items[second] = temp;

      if (this._items[first]) {
        this._items[first].slot = first;
      }
      if (this._items[second]) {
        this._items[second].slot = second;
      }

      globalGame.client.sendInventorySwap(first, second);
    },

    toArray: function () {
      return this._items;
    },

    loadFromObject: function (data) {
      if ($.isEmptyObject(data)) {
        return;
      }

      this._size = data[0];
      this.reset();

      var ids = {};
      $.each(data[1], function (id, item) {
        if (item) {
          this._items[item.slot] = InventoryItemFactory.getCreate(item.id, item);
          ids[item.id] = true;
        }
      }.bind(this));

      for (var id in globalInventoryItems) {
        if (!(id in ids)) {
          var inventoryItem = globalInventoryItems[id];
          inventoryItem.isDeleted = true;
          delete globalInventoryItems[id];
        }
      }

      this.update();
    },

    reset: function () {
      this._items = [];
      for (var i = 0; i < this.size; i++) {
        this._items[i] = null;
      }
    },

    serialize: function () {
      var items = [];
      for (var i in this.items) {
        if (!this.items[i]) {
          continue;
        }

        items[i] = this.items[i].getData();
      }

      return [this.size, items];
    },

    update: function () {
      globalGame.updateInventory();
    },
  });

  return Inventory;
});

File: client/dart/inventory.dart
library inventory;

import 'dart:collection';

import 'base.dart';
import 'game.dart';
import 'inventoryitem.dart';
import 'lib/gametypes.dart';

class Inventory extends Base {

  Map<int, InventoryItem> _items = {};
  int size = 12;

  Inventory(data) {
    if (data != null) {
      this.loadFromObject(data);
    }
  }

  InventoryItem find(String id) {
    return this._items.values.firstWhere(
      (item) => item.id == id,
      orElse: () => null
    );
  }

  void throwItem(int slot) {
    InventoryItem item = this._items[slot];
    if (item != null) {
      this.remove(slot);
      Game.client.sendThrowItem(item);
    }
  }

  /**
   * The inventory slot pointed to by source can never be null as that
   * should be the swap source.
   *
   * We are optimistically doing inventory swaps on the client side.
   */
  void swap(int sourceSlot, int destinationSlot) {
    InventoryItem source = this._items[sourceSlot];

    // don't fill the items array with nulls when the destination is an empty slot
    if (this._items.containsKey(destinationSlot)) {
      this._items[sourceSlot] = this._items[destinationSlot];
    } else {
      this._items.remove(sourceSlot);
    }

    this._items[destinationSlot] = source;

    Game.client.sendInventorySwap(sourceSlot, destinationSlot);
  }

  InventoryItem get(int slot) {
    return this._items[slot];
  }

  void remove(int slot) {
    this._items.remove(slot);
  }

  void reset() {
    this._items.clear();
  }

  void update() {
    Game.updateInventory();
  }

  List<dynamic> serialize() {
    return [
      this.size,
      this._items.values.map((InventoryItem item) => item.serialize()).toList()
    ];
  }

  void forEach(void callback(int, InventoryItem)) {
    Map<int, InventoryItem> items = new SplayTreeMap.from(this._items);

    // fill empty slots with null
    for (int i = 0; i < this.size; ++i) {
      if (!items.containsKey(i)) {
        items[i] = null;
      }
    }

    items.forEach(callback);
  }

  void loadFromObject(List<dynamic> data) {
    if (data == null || data.isEmpty) {
      return;
    }

    this.size = data[0];
    this.reset();

    data[1].forEach((String id, Map<String, dynamic> item) {
      this._items[item["slot"]] = new InventoryItem(Entities.get(item['kind']), item);
    });

    this.update();
  }
}
